ECC support for Key Vault SDK

msfcolombo opened the pull request:
<!-- DO NOT DELETE THIS TEMPLATE -->

## Description
- Added Elliptic Curve Cryptography fields and types to JsonWebKey classes.
- Added support to ECDSA algorithms to Key Vault core. The following curves are supported: `P-256`, `P-384`, `P-521` and `SECP256K1`.
- Miscellaneous cleanups and enhancements.
- [Merged PR with swagger spec changes](https://github.com/Azure/azure-rest-api-specs/pull/1538).
---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md).**
- [x] **The pull request does not introduce [breaking changes](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/Documentation/breaking-changes.md).**

### [General Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#sdk-generation-guidelines)
- [x] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [x] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code.
- [x] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK.

msfcolombo committed:
Official ECC support for Key Vault SDK

msftclas commented:
@msfcolombo,
Thanks for your contribution as a Microsoft full-time employee or intern. You do not need to sign a CLA.
_Thanks,
Microsoft Pull Request Bot_

shahabhijeet commented:
@msfcolombo 
This PR is good to go once someone approves from key vault team (e.g. Harvey/Rick Randal)

JeffSimmer commented on src/SDKs/KeyVault/dataPlane/Microsoft.Azure.KeyVault/Generated/KeyVaultClientExtensions.cs:
```diff
@@ -60,9 +63,9 @@ public static partial class KeyVaultClientExtensions
             /// The cancellation token.
             /// </param>
-            public static async Task<KeyBundle> CreateKeyAsync(this IKeyVaultClient operations, string vaultBaseUrl, string keyName, string kty, int? keySize = default(int?), IList<string> keyOps = default(IList<string>), KeyAttributes keyAttributes = default(KeyAttributes), IDictionary<string, string> tags = default(IDictionary<string, string>), CancellationToken cancellationToken = default(Canc ...
+            public static async Task<KeyBundle> CreateKeyAsync(this IKeyVaultClient operations, string vaultBaseUrl, string keyName, string kty, string curve = default(string), int? keySize = default(int?), IList<string> keyOps = default(IList<string>), KeyAttributes keyAttributes = default(KeyAttributes), IDictionary<string, string> tags = default(IDictionary<string, string>), CancellationToken  ...
```
This is a breaking API change for both code and existing binaries.  Do we need to make a breaking change for this?

msfcolombo replied:
A complete solution for this would be using a structure for key creation parameters. Otherwise every new key feature would cause a breaking change. But in order to avoid breaking this time, we would need Autorest to generate an overload. I will play with Swagger and Autorest to check if I can make it generate method overloads. If I can't find a solution quickly, I will change minor version.

msfcolombo replied:
@JeffSimmer I restored compatibility. Please take a look commit 296dbdb.

JeffSimmer replied:
Looks good, I think the NewKeyParameters class is much nicer from a versioning perspective.  Thanks!

msfcolombo committed:
Improvements to Key Vault ECC support and compatibility with previous versions.

-Restored order of create key parameters.
-Added a new method for creating keys which takes a structure as parameter and is more future-proof.
-Renamed a few fields and types to make more readable and intuitive.
-Added create/sign/verify tests cases for ECC keys.

shahabhijeet commented:
@msfcolombo need signoff from your team (Rich or Hervey)

msfcolombo committed:
Improved JsonWebKey in order to better accomodate protocol changes.

-JsonWebKey does not automatically validate itself during deserialization; this prevents cryptic or out-of-context error messages.
-JsonWebKey is now sealed and stores JSON properties that are not part of schema; this allows the class to support future changes to the protocol.
-Verification logic was moved to the new type JsonWebKeyVerifier, which provides rich features such as plugability for new key types.
-Fixed letter case on informal texts such as assembly title.
-Bumped some versions to reflect changes and compatibility breaks.

shahabhijeet commented:
@msfcolombo this PR will be closed by EOW. You can always reopen again.
This PR needs an approval from a KeyVault team members I have mentioned above.

msfcolombo commented:
Closing for now. Will reopen again after I get internal sign-off.
